### PR TITLE
AzureCreateVirtualMachineTask: introduce new task

### PR DESCRIPTION
**Pull request by nirarg**
Add new task creating VM on Azure cloud
Currently it create Ubuntu

**What this PR does / why we need it**:

This PR is created as part of the next demo requirements

**Which issue(s) this PR fixes*:
Fixes #[FLPATH-362](https://issues.redhat.com/browse/FLPATH-362)

**Change type**
- [x] New feature
- [ ] Bug fix
- [ ] Unit tests
- [ ] Integration tests
- [ ] CI
- [ ] Documentation
- [ ] Auto-generated SDK code

**Impacted services**
- [x] Workflow Service
- [ ] Notification Service

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.

### CODE REVIEW

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import com.redhat.parodos.workflows.work.WorkContext;
+import com.redhat.parodos.workflows.work.WorkReport;
+import com.redhat.parodos.workflows.work.WorkStatus;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+import org.springframework.util.StringUtils;
+
+@Slf4j
+public class AzureCreateVirtualMachineTask extends BaseInfrastructureWorkFlowTask {
```
Comment (masayag) [inline prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:35]: thinking if `Azure` prefix is needed for the class name if it is already placed in `.azure` package?

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
+package com.redhat.parodos.tasks.azure;
+
+import java.util.List;
+
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.http.policy.HttpLogDetailLevel;
+import com.azure.core.management.AzureEnvironment;
+import com.azure.core.management.Region;
+import com.azure.core.management.profile.AzureProfile;
+import com.azure.identity.*;
```
Comment (masayag) [inline prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:10]: pls expand.

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+					.create();
+
+			PublicIpAddress publicIpAddress = virtualMachine.getPrimaryPublicIPAddress();
+			if (publicIpAddress == null) {
+				log.error("VirtualMachine was created but without public IP");
+			}
+
+			String ipAddress = publicIpAddress.ipAddress();
+			log.info("VirtualMachine was created with public IP {}", ipAddress);
+			context.put("public-ip-address", ipAddress);
```
Comment (masayag) [inline prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:128]: Can this be described somewhere? Seems like a part of the task's output.

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+			VirtualMachine virtualMachine = azureResourceManager.virtualMachines().define(resourcesPrefix + "VM")
+					.withRegion(Region.US_EAST).withExistingResourceGroup(resourcesPrefix + "ResourceGroup")
+					.withExistingPrimaryNetworkInterface(networkInterface)
+					.withPopularLinuxImage(KnownLinuxVirtualMachineImage.UBUNTU_SERVER_18_04_LTS)
+					.withRootUsername(userName).withSsh(sshKey).withComputerName(resourcesPrefix + "VM")
+					.withExistingAvailabilitySet(availabilitySet).withSize(VirtualMachineSizeTypes.STANDARD_D3_V2)
+					.create();
+
+			PublicIpAddress publicIpAddress = virtualMachine.getPrimaryPublicIPAddress();
+			if (publicIpAddress == null) {
```
Comment (masayag) [inline prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:122]: Is it a simple getter? can the IP address be configured after the VM is actually up and running and obtained the address?

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+			NetworkInterface networkInterface = azureResourceManager.networkInterfaces().define(resourcesPrefix + "NIC")
+					.withRegion(Region.US_EAST).withExistingResourceGroup(resourcesPrefix + "ResourceGroup")
+					.withExistingPrimaryNetwork(network).withSubnet(resourcesPrefix + "Subnet")
+					.withPrimaryPrivateIPAddressDynamic().withExistingPrimaryPublicIPAddress(publicIPAddress).create();
+
+			log.info("Creating virtual machine...");
+			VirtualMachine virtualMachine = azureResourceManager.virtualMachines().define(resourcesPrefix + "VM")
+					.withRegion(Region.US_EAST).withExistingResourceGroup(resourcesPrefix + "ResourceGroup")
+					.withExistingPrimaryNetworkInterface(networkInterface)
+					.withPopularLinuxImage(KnownLinuxVirtualMachineImage.UBUNTU_SERVER_18_04_LTS)
```
Comment (RichardW98) [inline prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:116]: should we make this dynamic to support windows server or other systems?

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java (hunk 1 above)
Comment (nirarg) [reply prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:35]: This is the naming convention used by all tasks under Prebuild module
Do you think we should change it?

Comment (masayag) [reply prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:35]: don't have a strong opinion on either way.

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java (hunk 5 above)
Comment (masayag) [reply prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:116]: +1, I'd suggest to parameterized it (or a closed list of options) + also let the user specify VirtualMachineSizeTypes

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java (hunk 4 above)
Comment (nirarg) [reply prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:122]: The public IP is created before the Virtual machine created, therefore it shouldn't be empty at this point.
I added this validation just to be on the safe side

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java (hunk 5 above)
Comment (nirarg) [reply prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:116]: I suggest to have this change after the demo

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java (hunk 3 above)
Comment (nirarg) [reply prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:135]: Added description to the execute function, any other idea how to describe this?

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java (hunk 5 above)
Comment (nirarg) [reply prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:123]: As from the first impression I got, its not just about changing the machine type, also the function (withPopularLinuxImage) and the size might be changed .
Need to investigate more about the types included.

File: prebuilt-tasks/pom.xml [hunk 6]
```diff
@@ -134,5 +134,22 @@
             <artifactId>spring-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <!-- Start libraries for Azure access -->
+        <dependency>
+            <groupId>com.azure</groupId>
+            <artifactId>azure-core</artifactId>
+            <version>1.37.0</version>
```
Comment (pkliczewski) [inline prebuilt-tasks/pom.xml:141]: please use properties for versions like we do it for other dependencies.

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+			String ipAddress = publicIpAddress.ipAddress();
+			log.info("VirtualMachine was created with public IP {}", ipAddress);
+			context.put("public-ip-address", ipAddress);
+
+			return new DefaultWorkReport(WorkStatus.COMPLETED, context, null);
+		}
+		catch (MissingParameterException e) {
+			log.error("Task {} failed: missing required parameter, error: {}", getName(), e.getMessage());
+		}
+		catch (Exception e) {
```
Comment (pkliczewski) [inline prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:141]: please be specific which exceptions we want to catch here

**Review by pkliczewski — COMMENTED** [review]
please add some tests

**Comment by nirarg** [comment]
> please add some tests

@pkliczewski 
I'm working on adding unit tests for the new task
Maybe would add this in new PR, in order to have this PR merged before the demo

File: prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java (hunk 7 above)
Comment (nirarg) [reply prebuilt-tasks/src/main/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTask.java:141]: Except to `MissingParameterException` there is no known declared Exception that should be catched
Therefore I catch `Exception` to cover any unexpected `RuntimeException`

File: prebuilt-tasks/pom.xml (hunk 6 above)
Comment (nirarg) [reply prebuilt-tasks/pom.xml]: Done

**Review by pkliczewski — COMMENTED** [review]
/lgtm

**Comment by nirarg** [comment]
Added unit tests to the PR
Added all new changes in different commit: https://github.com/parodos-dev/parodos/pull/346/commits/887c63308521e5ebc45cdca87babd2228962fff2

File: prebuilt-tasks/src/test/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTaskTest.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+		when(azureResourceClient.createNetwork(resourcesPrefix)).thenReturn(network);
+		when(azureResourceClient.createNetworkInterface(resourcesPrefix, network, publicIPAddress))
+				.thenReturn(networkInterface);
+		when(azureResourceClient.createVirtualMachine(resourcesPrefix, networkInterface, availabilitySet, userName,
+				sshKey)).thenReturn(virtualMachine);
+	}
+
+	@Test
+	public void testHappyFlow() {
+		String publicIP = "11.11.11.11";
```
Comment (masayag) [inline prebuilt-tasks/src/test/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTaskTest.java:94]: pls add tests using the convention:
```
/ /given
...
// when
..
// then
..
```
see as en example https://github.com/parodos-dev/parodos/blob/main/workflow-service/src/test/java/com/redhat/parodos/workflow/execution/service/WorkFlowServiceImplTest.java#L175

Comment (nirarg) [reply prebuilt-tasks/src/test/java/com/redhat/parodos/tasks/azure/AzureCreateVirtualMachineTaskTest.java:95]: Done

**Comment by nirarg** [comment]
Also added instructions how to use the example workflow configuration
see [here](https://github.com/parodos-dev/parodos/pull/346/commits/463e04a2de9f25f7855f4977d0224ff5b400f9db)

**Pull request merged**